gateway/ln-gateway/src/bin/cln_extension.rs

```diff
@@ -559,59 +564,117 @@ impl GatewayLightning for ClnRpcService {
+                // failure in blocks
+                let delay = route
+                    .first()
+                    .ok_or_else(|| {
+                        tonic::Status::internal(format!(
+                            "Returned route did not have any hops for payment_hash: {payment_hash}"
+                        ))
+                    })?
+                    .delay;
+                if max_delay < delay.into() {
+                    return Err(tonic::Status::internal(format!("Worst case timeout for the payment is too long. max_delay: {max_delay} delay: {delay} payment_hash: {payment_hash}")));
+                }
 
-        let preimage = self
-            .pay_with_route(pruned_invoice, payment_hash, route)
-            .await
-            .map_err(|err| tonic::Status::internal(err.to_string()))?;
-        let response = PayInvoiceResponse { preimage };
+                // Verify the total fee is less than `max_fee_msat`
+                let first_hop_amount = route
+                    .first()
+                    .ok_or_else(|| {
+                        tonic::Status::internal(format!(
+                            "Returned route did not have any hops for payment_hash: {payment_hash}"
+                        ))
+                    })?
+                    .amount_msat;
+                let last_hop_amount = route
+                    .last()
+                    .ok_or_else(|| {
+                        tonic::Status::internal(format!(
+                            "Returned route did not have any hops for payment_hash: {payment_hash}"
+                        ))
+                    })?
+                    .amount_msat;
+                let fee = first_hop_amount - last_hop_amount;
+                if max_fee_msat < fee.msat() {
+                    return Err(tonic::Status::internal(format!(
+                        "Fee: {} for payment {payment_hash} is greater than max_fee_msat: {max_fee_msat}",
+                        fee.msat()
+                    )));
+                }
 
-        Ok(tonic::Response::new(response))
+                debug!(
+                    ?route_attempt,
+                    ?payment_hash,
+                    ?route,
+                    "Attempting payment with route"
+                );
+                match self
+                    .pay_with_route(pruned_invoice.clone(), payment_hash, route.clone())
+                    .await
+                {
+                    Ok(preimage) => {
+                        let response = PayInvoiceResponse { preimage };
+                        return Ok(tonic::Response::new(response));
+                    }
+                    Err(e) => {
+                        error!(
+                            ?route_attempt,
+                            ?payment_hash,
+                            ?e,
+                            "Pruned invoice payment attempt failure"
+                        );
```

Comment:
does this error contain any useful info about the node that failed? if you're a single channel to an LSP you don't want to ban your own lsp, should only exclude the specific along the route that failed.

Reply:
Fixed. Now only returns the erring node.